<a href="https://colab.research.google.com/github/tass-ouaer/Data-cleaning-Online_Courses_Dataset/blob/main/Online_course_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [677]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from google.colab import drive

In [678]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [679]:
df=pd.read_csv('/content/drive/MyDrive/Internship test/Online_Courses.csv')

In [680]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 45 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             8092 non-null   int64  
 1   Title                  8092 non-null   object 
 2   URL                    8092 non-null   object 
 3   Short Intro            8040 non-null   object 
 4   Category               2819 non-null   object 
 5   Sub-Category           2819 non-null   object 
 6   Course Type            2819 non-null   object 
 7   Language               2819 non-null   object 
 8   Subtitle Languages     2802 non-null   object 
 9   Skills                 2099 non-null   object 
 10  Instructors            2808 non-null   object 
 11  Rating                 2742 non-null   object 
 12  Number of viewers      2695 non-null   object 
 13  Duration               7944 non-null   object 
 14  Site                   8092 non-null   object 
 15  Prog

We notice that a lot of columns have many nulls (more than 60% of their records are null).
Also we have duplicate columns for the same data so we will check them and drop the duplicate columns or use them to replace nulls.

In [681]:
df.head()

Unnamed: 0                                              Title  \
0           0                    Machine Learning Specialization   
1           1        Introduction to Data Science Specialization   
2           2  Data Science Fundamentals with Python and SQL ...   
3           3       Key Technologies for Business Specialization   
4           4                       Deep Learning Specialization   

                                                 URL  \
0  https://www.coursera.org/specializations/machi...   
1  https://www.coursera.org/specializations/intro...   
2  https://www.coursera.org/specializations/data-...   
3  https://www.coursera.org/specializations/key-t...   
4  https://www.coursera.org/specializations/deep-...   

                                         Short Intro      Category  \
0  #BreakIntoAI with Machine Learning Specializat...  Data Science   
1  Launch your career in data science. Gain found...  Data Science   
2  Build the Foundation for your Data Science car...  Data Science   
3  Get Ahead with Key Business Technologies. Gain...      Business   
4  Become a Machine Learning expert. Master the f...  Data Science   

          Sub-Category     Course Type Language  \
0     Machine Learning  Specialization  English   
1        Data Analysis  Specialization  English   
2        Data Analysis  Specialization  English   
3  Business Essentials  Specialization  English   
4     Machine Learning  Specialization  English   

                                  Subtitle Languages  \
0                                 Subtitles: English   
1  Subtitles: English, Arabic, French, Portuguese...   
2  Subtitles: English, Arabic, French, Portuguese...   
3  Subtitles: English, Arabic, French, Portuguese...   
4  Subtitles: English, Chinese (Traditional), Ara...   

                                              Skills  ... Course Short Intro  \
0  Decision Trees, Artificial Neural Network, Log...  ...                NaN   
1  Data Science, Relational Database Management S...  ...                NaN   
2  Data Science, Github, Python Programming, Jupy...  ...                NaN   
3  Data Science, Artificial Intelligence (AI), Bu...  ...                NaN   
4  Artificial Neural Network, Convolutional Neura...  ...                NaN   

  Weekly study Premium course What's include Rank Created by Program  \
0          NaN            NaN            NaN  NaN        NaN     NaN   
1          NaN            NaN            NaN  NaN        NaN     NaN   
2          NaN            NaN            NaN  NaN        NaN     NaN   
3          NaN            NaN            NaN  NaN        NaN     NaN   
4          NaN            NaN            NaN  NaN        NaN     NaN   

  Number of ratings Price  COURSE CATEGORIES  
0               NaN   NaN                NaN  
1               NaN   NaN                NaN  
2               NaN   NaN                NaN  
3               NaN   NaN                NaN  
4               NaN   NaN                NaN  

[5 rows x 45 columns]

In [682]:
df.tail()

Unnamed: 0                                              Title  \
8087         143                             Certified Scrum Master   
8088         144             Salesforce Basics Course for Beginners   
8089         145  ICP-ACC (ICAgile Certified Agile Coaching) Cer...   
8090         146                                   /irisprodflip456   
8091         147                       Introduction to Industry 4.0   

                                                    URL  \
8087  https://www.simplilearn.com/certified-scrum-ma...   
8088  https://www.simplilearn.com/salesforce-course?...   
8089  https://www.simplilearn.com/certified-agile-co...   
8090   https://www.simplilearn.com/irisprodflip456?tag=   
8091  https://www.simplilearn.com/introduction-indus...   

                                            Short Intro Category Sub-Category  \
8087  This Certified ScrumMaster® (CSM) certificatio...      NaN          NaN   
8088                                                NaN      NaN          NaN   
8089  This ICP-ACC Certification Training Course aim...      NaN          NaN   
8090                          Iris Production check 456      NaN          NaN   
8091                      Introduction to Industry 4.0       NaN          NaN   

     Course Type Language Subtitle Languages Skills  ... Course Short Intro  \
8087         NaN      NaN                NaN    NaN  ...                NaN   
8088         NaN      NaN                NaN    NaN  ...                NaN   
8089         NaN      NaN                NaN    NaN  ...                NaN   
8090         NaN      NaN                NaN    NaN  ...                NaN   
8091         NaN      NaN                NaN    NaN  ...                NaN   

     Weekly study Premium course What's include Rank Created by Program  \
8087          NaN            NaN            NaN  NaN        NaN  Course   
8088          NaN            NaN            NaN  NaN        NaN  Course   
8089          NaN            NaN            NaN  NaN        NaN  Course   
8090          NaN            NaN            NaN  NaN        NaN  Course   
8091          NaN            NaN            NaN  NaN        NaN  Course   

     Number of ratings Price        COURSE CATEGORIES  
8087               369   NaN          Agile and Scrum  
8088               348   NaN                 Big Data  
8089               301   NaN          Agile and Scrum  
8090               293   NaN       Big Data Analytics  
8091               240   NaN  Business and Leadership  

[5 rows x 45 columns]

In [683]:
df = df.drop(columns=['Unnamed: 0'])

We will create new clean columns for the ones that are duplicated.
Even though Title and URL are duplicated,they have no nulls so will keep them and drop their duplicates.


In [684]:
# Create a new, clean 'Category' column
df['Master_Category'] = df['Category'].fillna(df['COURSE CATEGORIES'])

#For Short Intro
df['Master_Short_Intro'] = df['Short Intro'].fillna(df['Course Short Intro'])

In [685]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8092 entries, 0 to 8091
Data columns (total 46 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Title                  8092 non-null   object 
 1   URL                    8092 non-null   object 
 2   Short Intro            8040 non-null   object 
 3   Category               2819 non-null   object 
 4   Sub-Category           2819 non-null   object 
 5   Course Type            2819 non-null   object 
 6   Language               2819 non-null   object 
 7   Subtitle Languages     2802 non-null   object 
 8   Skills                 2099 non-null   object 
 9   Instructors            2808 non-null   object 
 10  Rating                 2742 non-null   object 
 11  Number of viewers      2695 non-null   object 
 12  Duration               7944 non-null   object 
 13  Site                   8092 non-null   object 
 14  Program Type           282 non-null    object 
 15  Cour

We notice that Master short Intro didn't improve much our column so we can keep the original column.
We will drop the other columns that are duplicated.

In [686]:
df = df.drop(columns=['Category','Course Title','Course URL','COURSE CATEGORIES','Short Intro','Course Short Intro','Master_Short_Intro','Unique Projects'])

In [687]:
df.duplicated().sum()

np.int64(5)

In [688]:
df=df.drop_duplicates()

In [689]:
df.isnull().sum()

Title                       0
URL                         0
Sub-Category             5268
Course Type              5268
Language                 5268
Subtitle Languages       5285
Skills                   5993
Instructors              5279
Rating                   5350
Number of viewers        5392
Duration                  148
Site                        0
Program Type             7810
Courses                  3148
Level                    7906
Number of Reviews        8045
Prequisites              7811
What you learn           7814
Related Programs         8007
Monthly access           7994
6-Month access           8085
4-Month access           8053
3-Month access           8062
5-Month access           8085
2-Month access           8077
School                   7821
Topics related to CRM    3421
ExpertTracks             3244
FAQs                     7004
Weekly study             3244
Premium course           6465
What's include           7668
Rank                     8050
Created by               8050
Program                  7939
Number of ratings        7961
Price                    8022
Master_Category          5120
dtype: int64

In [690]:
# Calculate the percentage of missing values for each column
missing_percentage = df.isnull().mean()

In [691]:
missing_percentage

Title                    0.000000
URL                      0.000000
Sub-Category             0.651416
Course Type              0.651416
Language                 0.651416
Subtitle Languages       0.653518
Skills                   0.741066
Instructors              0.652776
Rating                   0.661556
Number of viewers        0.666749
Duration                 0.018301
Site                     0.000000
Program Type             0.965747
Courses                  0.389267
Level                    0.977618
Number of Reviews        0.994806
Prequisites              0.965871
What you learn           0.966242
Related Programs         0.990108
Monthly access           0.988500
6-Month access           0.999753
4-Month access           0.995796
3-Month access           0.996909
5-Month access           0.999753
2-Month access           0.998763
School                   0.967108
Topics related to CRM    0.423025
ExpertTracks             0.401138
FAQs                     0.866081
Weekly study             0.401138
Premium course           0.799431
What's include           0.948188
Rank                     0.995425
Created by               0.995425
Program                  0.981699
Number of ratings        0.984419
Price                    0.991962
Master_Category          0.633115
dtype: float64

We will focus our efforts only on the columns needed for the dashboard.

We will drop columns that will not be used later in the analysis nor in the dashboard that also have a high percentage of nulls.


In [692]:
df=df.drop(columns=['Monthly access','6-Month access','4-Month access','3-Month access','5-Month access','2-Month access','Rank','Prequisites','What you learn','Related Programs','School','FAQs','What\'s include','Created by','Program','Courses','ExpertTracks','Weekly study','Topics related to CRM','Sub-Category','Subtitle Languages','Premium course'])

In [693]:
#recalcultae the new missing percentages
missing_percentage = df.isnull().mean()
missing_percentage.sort_values()

Title                0.000000
URL                  0.000000
Site                 0.000000
Duration             0.018301
Master_Category      0.633115
Course Type          0.651416
Language             0.651416
Instructors          0.652776
Rating               0.661556
Number of viewers    0.666749
Skills               0.741066
Program Type         0.965747
Level                0.977618
Number of ratings    0.984419
Price                0.991962
Number of Reviews    0.994806
dtype: float64

Several columns are unusable. If a column is over 90% empty, it is statistically invalid for analysis.
We will drop these columns.

Price (99.19%)

Level (97.76%)

Program Type (96.57%)

Number of ratings (98.44%)

Number of Reviews (99.48%)



In [694]:
df=df.drop(columns=(['Price','Level','Program Type','Number of ratings','Number of Reviews']))

In [695]:
#recalcultae the new missing percentages
missing_percentage = df.isnull().mean()
missing_percentage.sort_values()

Title                0.000000
URL                  0.000000
Site                 0.000000
Duration             0.018301
Master_Category      0.633115
Course Type          0.651416
Language             0.651416
Instructors          0.652776
Rating               0.661556
Number of viewers    0.666749
Skills               0.741066
dtype: float64

Now we are left with the columns that we NEED and CAN use for our dashboard.



Even though the instructions says Price needs to be included in the dashboard later,we must make data driven decisions and in this case drop the column price and remove the following from our dashboard plan:

KPI: Median Price

KPI: % Free Courses

Chart: Price vs Rating scatter plot

We will Start the cleaning of our Columns

In [696]:
#distinct values of each column
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Distinct values in column '{column}':")
    print(unique_values)
    print()

Distinct values in column 'Title':
['Machine Learning Specialization'
 'Introduction to Data Science Specialization'
 'Data Science Fundamentals with Python and SQL Specialization' ...
 'ICP-ACC (ICAgile Certified Agile Coaching) Certification Training Course'
 '/irisprodflip456' 'Introduction to Industry 4.0']

Distinct values in column 'URL':
['https://www.coursera.org/specializations/machine-learning-introduction'
 'https://www.coursera.org/specializations/introduction-data-science'
 'https://www.coursera.org/specializations/data-science-fundamentals-python-sql'
 ...
 'https://www.simplilearn.com/certified-agile-coaching-icp-acc-certification-training?tag='
 'https://www.simplilearn.com/irisprodflip456?tag='
 'https://www.simplilearn.com/introduction-industry-4-bcg?tag=']

Distinct values in column 'Course Type':
['Specialization' 'Professional Certificate' 'Course' 'Project' nan]

Distinct values in column 'Language':
['English' 'Spanish' 'Portuguese (Brazilian)' 'French' 'Japanese

In [697]:
df.head()

Title  \
0                    Machine Learning Specialization   
1        Introduction to Data Science Specialization   
2  Data Science Fundamentals with Python and SQL ...   
3       Key Technologies for Business Specialization   
4                       Deep Learning Specialization   

                                                 URL     Course Type Language  \
0  https://www.coursera.org/specializations/machi...  Specialization  English   
1  https://www.coursera.org/specializations/intro...  Specialization  English   
2  https://www.coursera.org/specializations/data-...  Specialization  English   
3  https://www.coursera.org/specializations/key-t...  Specialization  English   
4  https://www.coursera.org/specializations/deep-...  Specialization  English   

                                              Skills  \
0  Decision Trees, Artificial Neural Network, Log...   
1  Data Science, Relational Database Management S...   
2  Data Science, Github, Python Programming, Jupy...   
3  Data Science, Artificial Intelligence (AI), Bu...   
4  Artificial Neural Network, Convolutional Neura...   

                                         Instructors    Rating  \
0  Andrew Ng ,Eddy Shyu ,Aarti Bagul ,Geoff Ladwig ,  4.9stars   
1  Rav Ahuja ,Alex Aklson ,Aije Egwaikhide ,Svetl...  4.7stars   
2  Aije Egwaikhide ,Svetlana Levitan ,Romeo Kienz...  4.6stars   
3                           Rav Ahuja ,Alex Aklson ,  4.7stars   
4  Andrew Ng ,Kian Katanforoosh ,Younes Bensouda ...  4.9stars   

  Number of viewers                            Duration      Site  \
0           10,438   Approximately 3 months to complete  Coursera   
1           11,927   Approximately 5 months to complete  Coursera   
2            2,295   Approximately 7 months to complete  Coursera   
3            1,232   Approximately 3 months to complete  Coursera   
4          129,253   Approximately 5 months to complete  Coursera   

  Master_Category  
0    Data Science  
1    Data Science  
2    Data Science  
3        Business  
4    Data Science

In [698]:
df.tail()

Title  \
8087                             Certified Scrum Master   
8088             Salesforce Basics Course for Beginners   
8089  ICP-ACC (ICAgile Certified Agile Coaching) Cer...   
8090                                   /irisprodflip456   
8091                       Introduction to Industry 4.0   

                                                    URL Course Type Language  \
8087  https://www.simplilearn.com/certified-scrum-ma...         NaN      NaN   
8088  https://www.simplilearn.com/salesforce-course?...         NaN      NaN   
8089  https://www.simplilearn.com/certified-agile-co...         NaN      NaN   
8090   https://www.simplilearn.com/irisprodflip456?tag=         NaN      NaN   
8091  https://www.simplilearn.com/introduction-indus...         NaN      NaN   

     Skills Instructors Rating Number of viewers Duration         Site  \
8087    NaN         NaN    NaN               NaN      NaN  Simplilearn   
8088    NaN         NaN    NaN               NaN      NaN  Simplilearn   
8089    NaN         NaN    NaN               NaN      NaN  Simplilearn   
8090    NaN         NaN    NaN               NaN      NaN  Simplilearn   
8091    NaN         NaN    NaN               NaN      NaN  Simplilearn   

              Master_Category  
8087          Agile and Scrum  
8088                 Big Data  
8089          Agile and Scrum  
8090       Big Data Analytics  
8091  Business and Leadership

#Rating

In [699]:
df['Rating'] = df['Rating'].astype(str)

# Remove "stars"
df['Rating'] = df['Rating'].str.replace('stars', '', regex=False)
#clean whitespaces
df['Rating'] = df['Rating'].str.strip()
# Convert to numeric, coercing errors to NaN
df['Rating'] = pd.to_numeric(df['Rating'], errors='coerce')

# Display the first few rows and the data type of the cleaned column
print("Rating column after cleaning:")
df[['Rating']].head()

Rating column after cleaning:


Rating
0     4.9
1     4.7
2     4.6
3     4.7
4     4.9

To clean the nulls for this column we have 2 options:

Either drop the rows with null ratings or do a grped imputation based on course type and category(replace null with the average rating for each course type and category).

But since we have many nulls here and we will be using the average rating and top categories by rating, if we impute, we are no longer analyzing the data; we are just confirming the average we invented.

-> The final analysis will be based on truthful data. It is infinitely better to have a 100% accurate analysis of 2,700 courses than a 70% manufactured analysis of 8,000 courses.


In [700]:
#drop rows where rating is null
df=df.dropna(subset=['Rating'])

In [701]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2737 entries, 0 to 3074
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              2737 non-null   object 
 1   URL                2737 non-null   object 
 2   Course Type        2695 non-null   object 
 3   Language           2695 non-null   object 
 4   Skills             1945 non-null   object 
 5   Instructors        2694 non-null   object 
 6   Rating             2737 non-null   float64
 7   Number of viewers  2695 non-null   object 
 8   Duration           2737 non-null   object 
 9   Site               2737 non-null   object 
 10  Master_Category    2695 non-null   object 
dtypes: float64(1), object(10)
memory usage: 256.6+ KB


Our analysis will be based on 2737 rows.

In [702]:
#recalcultae the new missing percentages
missing_percentage = df.isnull().mean()
missing_percentage.sort_values()

Title                0.000000
URL                  0.000000
Rating               0.000000
Duration             0.000000
Site                 0.000000
Number of viewers    0.015345
Language             0.015345
Course Type          0.015345
Master_Category      0.015345
Instructors          0.015711
Skills               0.289368
dtype: float64

#Duration

In [703]:
#check distinct values in Duration
df['Duration'].unique()

array(['Approximately 3 months to complete',
       'Approximately 5 months to complete',
       'Approximately 7 months to complete',
       'Approximately 2 months to complete',
       'Approximately 4 months to complete',
       'Approximately 6 months to complete',
       'Approximately 8 months to complete',
       'Approximately 11 months to complete',
       'Approximately 10 months to complete',
       'Approximately 9 months to complete',
       'Approximately 12 months to complete',
       'Approximately 1 month to complete', '6 months at 10 hours a week',
       '5 months at 3 hours a week', '4 months at 10 hours a week',
       '3 months at 10 hours a week', '5 months at 10 hours a week',
       '12 months at 10 hours a week', '6 months at 3 hours a week',
       '7 months at 6 hours a week', '2 months at 10 hours a week',
       '8 months at 3 hours a week', '7 months at 2 hours a week',
       '4 months at 4 hours a week', '4 months at 5 hours a week',
       '8 months at

We notice a lot of formats in the rows. We need this in a single unit: Hour.

In [704]:
import re

def clean_duration(duration):
    if pd.isna(duration):
        return None

    # Normalization
    duration = duration.lower()
    duration = duration.replace('horas', 'hour').replace('meses', 'month').replace('semanas', 'week')
    duration = re.sub(r'[\n]', '', duration)
    duration = re.sub(r'\s+', ' ', duration).strip()
    duration = duration.replace('approx.', '').replace('approximately', '').replace('estimated', '').strip()

    # Pattern 1: "X months at Y hours a week"
    match = re.search(r'(\d+)\s*months?\s*at\s*(\d+)\s*hours?\s*a\s*week', duration)
    if match:
        months = int(match.group(1))
        hours_per_week = int(match.group(2))
        return months * 4 * hours_per_week

    # Pattern 2: "X - Y Weeks"
    match = re.search(r'(\d+)\s*-\s*(\d+)\s*weeks?', duration)
    if match:
        weeks_start = int(match.group(1))
        weeks_end = int(match.group(2))
        return ((weeks_start + weeks_end) / 2) * 5 # Assume 5 hours per week

    # Pattern 3: Simple Units ("X months", "X weeks", "X hours", "X minutes")
    match_month = re.search(r'(\d+)\s*months?', duration)
    if match_month:
        months = int(match_month.group(1))
        return months * 4 * 5 # Assume 4 weeks per month and 5 hours per week

    match_week = re.search(r'(\d+)\s*weeks?', duration)
    if match_week:
        weeks = int(match_week.group(1))
        return weeks * 5 # Assume 5 hours per week

    match_hour = re.search(r'(\d+)\s*hours?', duration)
    if match_hour:
        hours = int(match_hour.group(1))
        return hours

    match_minute = re.search(r'(\d+)\s*minutes?', duration)
    if match_minute:
        minutes = int(match_minute.group(1))
        return minutes / 60

    # Handle cases like "X Hour"
    match_single_hour = re.search(r'(\d+)\s*hour', duration)
    if match_single_hour:
        hours = int(match_single_hour.group(1))
        return hours

    return None

# Apply the cleaning function to create the new 'Duration_Hours' column
df['Duration_Hours'] = df['Duration'].apply(clean_duration)

# Display the first few rows with the new column and check info
print("DataFrame with new 'Duration_Hours' column:")
display(df[['Duration', 'Duration_Hours']].head())

print("\nInfo of the DataFrame with 'Duration_Hours':")
df.info()

DataFrame with new 'Duration_Hours' column:


Duration  Duration_Hours
0  Approximately 3 months to complete            60.0
1  Approximately 5 months to complete           100.0
2  Approximately 7 months to complete           140.0
3  Approximately 3 months to complete            60.0
4  Approximately 5 months to complete           100.0


Info of the DataFrame with 'Duration_Hours':
<class 'pandas.core.frame.DataFrame'>
Index: 2737 entries, 0 to 3074
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              2737 non-null   object 
 1   URL                2737 non-null   object 
 2   Course Type        2695 non-null   object 
 3   Language           2695 non-null   object 
 4   Skills             1945 non-null   object 
 5   Instructors        2694 non-null   object 
 6   Rating             2737 non-null   float64
 7   Number of viewers  2695 non-null   object 
 8   Duration           2737 non-null   object 
 9   Site               2737 non-null   object 
 10  Master_Category    2695 non-null   object 
 11  Duration_Hours     2737 non-null   float64
dtypes: float64(2), object(10)
memory usage: 278.0+ KB


In [705]:
df=df.drop(columns=['Duration'])

#Number of viewers

In [706]:
df['Number of viewers'].unique()

array(['10,438 ', '11,927 ', '2,295 ', ..., '2,810 ', '4,261 ', nan],
      dtype=object)

In [707]:
# Remove commas from 'Number of viewers' and then convert to float
df['Number of viewers'] = df['Number of viewers'].astype(str).str.replace(',', '', regex=False)
df['Number of viewers'] = pd.to_numeric(df['Number of viewers'], errors='coerce')

print("Number of viewers column after cleaning:")
df[['Number of viewers']].head()

Number of viewers column after cleaning:


Number of viewers
0            10438.0
1            11927.0
2             2295.0
3             1232.0
4           129253.0

In [708]:
#replace nulls by 0
df['Number of viewers'] = df['Number of viewers'].fillna(0)


#Course type

In [709]:
#check the first rows for each course type
df.groupby('Course Type').head(3)

Title  \
0                      Machine Learning Specialization   
1          Introduction to Data Science Specialization   
2    Data Science Fundamentals with Python and SQL ...   
414     Google Data Analytics Professional Certificate   
415  Google Project Management: Professional Certif...   
416          IBM Data Science Professional Certificate   
502  Become a CBRS Certified Professional Installer...   
503                         Introduction to Statistics   
504           Diversity and inclusion in the workplace   
992  ChatGPT for Beginners: Using AI For Market Res...   
993              Getting Started with Power BI Desktop   
996                 Curso Completo de Power BI Desktop   

                                                   URL  \
0    https://www.coursera.org/specializations/machi...   
1    https://www.coursera.org/specializations/intro...   
2    https://www.coursera.org/specializations/data-...   
414  https://www.coursera.org/professional-certific...   
415  https://www.coursera.org/professional-certific...   
416  https://www.coursera.org/professional-certific...   
502  https://www.coursera.org/learn/google-cbrs-cpi...   
503  https://www.coursera.org/learn/stanford-statis...   
504  https://www.coursera.org/learn/diversity-inclu...   
992  https://www.coursera.org/projects/chat-gpt-for...   
993  https://www.coursera.org/projects/power-bi-des...   
996  https://www.coursera.org/projects/curso-comple...   

                  Course Type Language  \
0              Specialization  English   
1              Specialization  English   
2              Specialization  English   
414  Professional Certificate  English   
415  Professional Certificate  English   
416  Professional Certificate  English   
502                    Course  English   
503                    Course  English   
504                    Course  English   
992                   Project  English   
993                   Project  English   
996                   Project  Spanish   

                                                Skills  \
0    Decision Trees, Artificial Neural Network, Log...   
1    Data Science, Relational Database Management S...   
2    Data Science, Github, Python Programming, Jupy...   
414  Spreadsheet,Data Cleansing,Data Analysis,Data ...   
415  Organizational Culture,Career Development,Stra...   
416  Data Science,Deep Learning,Machine Learning,Bi...   
502                                                NaN   
503                                                NaN   
504                                                NaN   
992  Business Analytics,Market Analysis,Business St...   
993  Business Analytics,Data Analysis,Data Visualiz...   
996                                                NaN   

                                           Instructors  Rating  \
0    Andrew Ng ,Eddy Shyu ,Aarti Bagul ,Geoff Ladwig ,     4.9   
1    Rav Ahuja ,Alex Aklson ,Aije Egwaikhide ,Svetl...     4.7   
2    Aije Egwaikhide ,Svetlana Levitan ,Romeo Kienz...     4.6   
414                         Google Career Certificates     4.8   
415                         Google Career Certificates     4.9   
416                                          Rav Ahuja     4.6   
502                                             Google     4.6   
503                                   Guenther Walther     4.6   
504                                       Junko TAKAGI     4.7   
992                                       Jan Demiralp     4.7   
993                                         Amit Yadav     4.7   
996                                        Leire Ahedo     4.4   

     Number of viewers      Site         Master_Category  Duration_Hours  
0              10438.0  Coursera            Data Science            60.0  
1              11927.0  Coursera            Data Science           100.0  
2               2295.0  Coursera            Data Science           140.0  
414                0.0  Coursera            Data Science           240.0  
415

We can see that there is a pattern:

titles that have specialization in their names are of course type specialization.

Titles that have Professional certificate are of course type Professional certificate

course and project types have no really remarkable pattern.

In [710]:
#for the nulls if title has Specialization then Course Type is specialization else if Professional Certificate then Professional Certificate else not specified
df['Course Type'] = df.apply(lambda row: 'Specialization' if 'Specialization' in row['Title'] else ('Professional Certificate' if 'Professional Certificate' in row['Title'] else 'Not Specified'), axis=1)

In [711]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 2737 entries, 0 to 3074
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              2737 non-null   object 
 1   URL                2737 non-null   object 
 2   Course Type        2737 non-null   object 
 3   Language           2695 non-null   object 
 4   Skills             1945 non-null   object 
 5   Instructors        2694 non-null   object 
 6   Rating             2737 non-null   float64
 7   Number of viewers  2737 non-null   float64
 8   Site               2737 non-null   object 
 9   Master_Category    2695 non-null   object 
 10  Duration_Hours     2737 non-null   float64
dtypes: float64(3), object(8)
memory usage: 256.6+ KB


#Language

In [712]:
df['Language'].unique()

array(['English', 'Spanish', 'Portuguese (Brazilian)', 'French',
       'Japanese', 'Arabic', 'German', 'Indonesian', 'Russian',
       'Chinese (Simplified)', nan], dtype=object)

In [713]:
df.groupby('Language').head(3)

Title  \
0                       Machine Learning Specialization   
1           Introduction to Data Science Specialization   
2     Data Science Fundamentals with Python and SQL ...   
443   Análisis de Datos de Google Professional Certi...   
445   Gestión de Proyectos de Google Professional Ce...   
449   Diseño de Experiencia del Usuario (UX) de Goog...   
453   Google Data Analytics (PT) Professional Certif...   
463      Google UX Design (PT) Professional Certificate   
464   Google Project Management (PT) Professional Ce...   
466      Google Data Analytics Professional Certificate   
467   Google Project Management Professional Certifi...   
468          Google データアナリティクス Professional Certificate   
470   تحليلات البيانات من Google Professional Certif...   
472           Google UX Design Professional Certificate   
474   Google IT Support (Deutsch)  Professional Cert...   
476   Google Project Management (DE) Professional Ce...   
478   Preparing for Google Cloud Certification: Clou...   
479       Analitik Data Google Professional Certificate   
480   Google Data Analytics (DE) Professional Certif...   
537                AI For Everyone （すべての人のためのAIリテラシー講座）   
1295                             Career Success Project   
1812                                       市场营销概论 (中文版）   
2000                                        财务会计概论（中文版）   
2177                                        企业金融概论（中文版）   

                                                    URL  \
0     https://www.coursera.org/specializations/machi...   
1     https://www.coursera.org/specializations/intro...   
2     https://www.coursera.org/specializations/data-...   
443   https://www.coursera.org/professional-certific...   
445   https://www.coursera.org/professional-certific...   
449   https://www.coursera.org/professional-certific...   
453   https://www.coursera.org/professional-certific...   
463   https://www.coursera.org/professional-certific...   
464   https://www.coursera.org/professional-certific...   
466   https://www.coursera.org/professional-certific...   
467   https://www.coursera.org/professional-certific...   
468   https://www.coursera.org/professional-certific...   
470   https://www.coursera.org/professional-certific...   
472   https://www.coursera.org/professional-certific...   
474   https://www.coursera.org/professional-certific...   
476   https://www.coursera.org/professional-certific...   
478   https://www.coursera.org/professional-certific...   
479   https://www.coursera.org/professional-certific...   
480   https://www.coursera.org/professional-certific...   
537   https://www.coursera.org/learn/ai-for-everyone-ja   
1295  https://www.coursera.org/learn/career-success-...   
1812  https://www.coursera.org/learn/wharton-shichan...   
2000  https://www.coursera.org/learn/wharton-kuaiji?...   
2177  https://www.coursera.org/learn/wharton-qiye-ji...   

                   Course Type                Language  \
0               Specialization                 English   
1               Specialization                 English   
2               Specialization                 English   
443   Professional Certificate                 Spanish   
445   Professional Certificate                 Spanish   
449   Professional Certificate                 Spanish   
453   Professional Certificate  Portuguese (Brazilian)   
463   Professional Certificate  Portuguese (Brazilian)   
464   Professional Certificate  Portuguese (Brazilian)   
466   Professional Certificate                  French   
467   Professional Certificate                  French   
468   Professional Certificate                Japanese   
470   Professional Certificate                  Arabic   
472   Professional Certificate                  French   
474   Professional Certificate                  German   
476   Professional Certificate                  German   
478   Professional Certificate                Japanese   
479   Professional Certificate              Indonesian

we can see that the titles that have a language diffrent than english in their names are in another language.

We can assume that the language of the tiitle is the language (there are some exceptions but we will assume that for simplicity)

In [714]:
#show rows with nulls
df[df['Language'].isnull()]

Title  \
2996               Artificial Intelligence for Trading   
3001                     How to Become a Data Engineer   
3002                                         Learn C++   
3003                                     Learn to Code   
3004                      How to Become a Data Analyst   
3005                  How to Become a Business Analyst   
3011                   How to Become a Product Manager   
3012          Programming for Data Science with Python   
3013                        AI Programming with Python   
3014                    How to Become a Data Scientist   
3015                           Front End Web Developer   
3019                          Full Stack Web Developer   
3020                  How to Become a Digital Marketer   
3022                                         Learn SQL   
3023                                       Learn React   
3024                                     Deep Learning   
3025                       How to Become a UX Designer   
3027       Become a Deep Reinforcement Learning Expert   
3028                   Become a Computer Vision Expert   
3029                       Become a Java Web Developer   
3030       Become a Natural Language Processing Expert   
3034                    Become a Cloud DevOps Engineer   
3036                    How to Become an iOS Developer   
3037              How to Become a Blockchain Developer   
3040                     Become an AWS Cloud Architect   
3041                 How to Become a Robotics Engineer   
3043                      Become an AI Product Manager   
3044              Learn Data Structures and Algorithms   
3047                          Become a Cloud Developer   
3048                   Become a Sensor Fusion Engineer   
3049                           AI for Business Leaders   
3050  Expand Your Knowledge of Artificial Intelligence   
3051                                   Data Streaming    
3054                                 AI for Healthcare   
3055                        Intro to Self-Driving Cars   
3056                        Become a Marketing Analyst   
3057            Intro to Machine Learning with PyTorch   
3058                 Predictive Analytics for Business   
3059         Intro to Machine Learning with TensorFlow   
3070         Flying Car and Autonomous Flight Engineer   
3072                                Data Visualization   
3074               Programming for Data Science with R   

                                                    URL    Course Type  \
2996  https://www.udacity.com/course/ai-for-trading-...  Not Specified   
3001  https://www.udacity.com/course/data-engineer-n...  Not Specified   
3002  https://www.udacity.com/course/c-plus-plus-nan...  Not Specified   
3003  https://www.udacity.com/course/intro-to-progra...  Not Specified   
3004  https://www.udacity.com/course/data-analyst-na...  Not Specified   
3005  https://www.udacity.com/course/business-analyt...  Not Specified   
3011  https://www.udacity.com/course/product-manager...  Not Specified   
3012  https://www.udacity.com/course/programming-for...  Not Specified   
3013  https://www.udacity.com/course/ai-programming-...  Not Specified   
3014  https://www.udacity.com/course/data-scientist-...  Not Specified   
3015  https://www.udacity.com/course/front-end-web-d...  Not Specified   
3019  https://www.udacity.com/course/full-stack-web-...  Not Specified   
3020  https://www.udacity.com/course/digital-marketi...  Not Specified   
3022    https://www.udacity.com/course/learn-sql--nd072  Not Specified   
3023  https://www.udacity.com/course/react-nanodegre...  Not Specified   
3024  https://www.udacity.com/course/deep-learning-n...  Not Specified   
3025  https://www.udacity.com/course/ux-designer-nan...  Not Specified   
3027  https://www.udacity.com/course/deep-reinforcem...  Not Specified   
3028  https://www.udacity.com/course/computer-vision...  Not Specified   
3029  https://www.udacity.com/course/java-developer-...  Not Specified   


In [715]:
#replace nulls by english
df['Language'] = df['Language'].fillna('English')

#Instructors

In [716]:
#replace nulls by unkown
df['Instructors'] = df['Instructors'].fillna('Unknown')

#Master Ctageory

In [717]:
df['Master_Category'].unique()

array(['Data Science', 'Business', 'Information Technology',
       'Computer Science', 'Arts and Humanities', 'Language Learning',
       'Personal Development', 'Social Sciences', 'Health',
       'Physical Science and Engineering', 'Math and Logic', '计算机科学',
       'Ciencia de Datos', 'Negocios', 'Ciencias de la Computación',
       'Negócios', 'データサイエンス', 'Tecnologia da informação', nan],
      dtype=object)

We clean the categories by mapping tables.

One map cleans the categories we already have. and the second maps key words in title to correct categories

In [718]:
translation_map = {
    # Spanish
    'ciencia de datos': 'Data Science',
    'negocios': 'Business',
    'ciencias de la computación': 'Computer Science',

    # Portuguese
    'negócios': 'Business',
    'tecnologia da informação': 'Information Technology',

    # Japanese
    'データサイエンス': 'Data Science',

    # Chinese
    '计算机科学': 'Computer Science'
}



In [719]:
df.groupby('Master_Category').head(3)

Title  \
0                      Machine Learning Specialization   
1          Introduction to Data Science Specialization   
2    Data Science Fundamentals with Python and SQL ...   
3         Key Technologies for Business Specialization   
9          Data Engineering Foundations Specialization   
10    IT Fundamentals for Cybersecurity Specialization   
11   Applied Software Engineering Fundamentals Spec...   
12   DevOps, Cloud, and Agile Foundations Specializ...   
13   Cloud Application Development Foundations Spec...   
14                 Python for Everybody Specialization   
18                 Business Foundations Specialization   
21   Introduction to Finance and Accounting Special...   
33                       Graphic Design Specialization   
35   Improve Your English Communication Skills Spec...   
61   Achieving Personal and Professional Success Sp...   
63   TESOL Certificate, Part 1: Teach English Now! ...   
65                       UI / UX Design Specialization   
68                     Creative Writing Specialization   
73   Methods and Statistics in Social Sciences Spec...   
84          The Business of Health Care Specialization   
112           Academic English: Writing Specialization   
119             Dynamic Public Speaking Specialization   
123        Music Education for Teachers Specialization   
126  Good with Words: Writing and Editing Specializ...   
127  Drug Development Product Management Specializa...   
129  Food Sustainability, Mindful Eating, and Healt...   
140  Algorithms for Battery Management Systems Spec...   
161    FPGA Design for Embedded Systems Specialization   
179  TESOL Certificate, Part 2: Teach English Now! ...   
189                    Renewable Energy Specialization   
208  Expressway to Data Science: Essential Math Spe...   
289  Advanced Statistics for Data Science Specializ...   
307           Mathematics for Engineers Specialization   
429  Meta Front-End Developer Professional Certificate   
443  Análisis de Datos de Google Professional Certi...   
445  Gestión de Proyectos de Google Professional Ce...   
449  Diseño de Experiencia del Usuario (UX) de Goog...   
455  Análise de marketing da Meta Professional Cert...   
468         Google データアナリティクス Professional Certificate   
487   Suporte em TI do Google Professional Certificate   

                                                   URL  \
0    https://www.coursera.org/specializations/machi...   
1    https://www.coursera.org/specializations/intro...   
2    https://www.coursera.org/specializations/data-...   
3    https://www.coursera.org/specializations/key-t...   
9    https://www.coursera.org/specializations/data-...   
10   https://www.coursera.org/specializations/it-fu...   
11   https://www.coursera.org/specializations/softw...   
12   https://www.coursera.org/specializations/devop...   
13   https://www.coursera.org/specializations/ibm-c...   
14     https://www.coursera.org/specializations/python   
18   https://www.coursera.org/specializations/whart...   
21   https://www.coursera.org/specializations/finan...   
33   https://www.coursera.org/specializations/graph...   
35   https://www.coursera.org/specializations/impro...   
61   https://www.coursera.org/specializations/whart...   
63      https://www.coursera.org/specializations/tesol   
65   https://www.coursera.org/specializations/ui-ux...   
68   https://www.coursera.org/specializations/creat...   
73   https://www.coursera.org/specializations/socia...   
84   https://www.coursera.org/specializations/healt...   
112  https://www.coursera.org/specializations/acade...   
119  https://www.coursera.org/specializations/publi...   
123  https://www.coursera.org/specializations/music...   
126  https://www.coursera.org/specializations/good-...   
127  https://www.coursera.org/specializations/drug-...   
129  https://www.coursera.org/specializations/food-...   
140  https://www.coursera.org/specializations/algor...   
161  https://www.coursera.org/specializations/

In [720]:
df[df['Master_Category'].isnull()]

Title  \
2996               Artificial Intelligence for Trading   
3001                     How to Become a Data Engineer   
3002                                         Learn C++   
3003                                     Learn to Code   
3004                      How to Become a Data Analyst   
3005                  How to Become a Business Analyst   
3011                   How to Become a Product Manager   
3012          Programming for Data Science with Python   
3013                        AI Programming with Python   
3014                    How to Become a Data Scientist   
3015                           Front End Web Developer   
3019                          Full Stack Web Developer   
3020                  How to Become a Digital Marketer   
3022                                         Learn SQL   
3023                                       Learn React   
3024                                     Deep Learning   
3025                       How to Become a UX Designer   
3027       Become a Deep Reinforcement Learning Expert   
3028                   Become a Computer Vision Expert   
3029                       Become a Java Web Developer   
3030       Become a Natural Language Processing Expert   
3034                    Become a Cloud DevOps Engineer   
3036                    How to Become an iOS Developer   
3037              How to Become a Blockchain Developer   
3040                     Become an AWS Cloud Architect   
3041                 How to Become a Robotics Engineer   
3043                      Become an AI Product Manager   
3044              Learn Data Structures and Algorithms   
3047                          Become a Cloud Developer   
3048                   Become a Sensor Fusion Engineer   
3049                           AI for Business Leaders   
3050  Expand Your Knowledge of Artificial Intelligence   
3051                                   Data Streaming    
3054                                 AI for Healthcare   
3055                        Intro to Self-Driving Cars   
3056                        Become a Marketing Analyst   
3057            Intro to Machine Learning with PyTorch   
3058                 Predictive Analytics for Business   
3059         Intro to Machine Learning with TensorFlow   
3070         Flying Car and Autonomous Flight Engineer   
3072                                Data Visualization   
3074               Programming for Data Science with R   

                                                    URL    Course Type  \
2996  https://www.udacity.com/course/ai-for-trading-...  Not Specified   
3001  https://www.udacity.com/course/data-engineer-n...  Not Specified   
3002  https://www.udacity.com/course/c-plus-plus-nan...  Not Specified   
3003  https://www.udacity.com/course/intro-to-progra...  Not Specified   
3004  https://www.udacity.com/course/data-analyst-na...  Not Specified   
3005  https://www.udacity.com/course/business-analyt...  Not Specified   
3011  https://www.udacity.com/course/product-manager...  Not Specified   
3012  https://www.udacity.com/course/programming-for...  Not Specified   
3013  https://www.udacity.com/course/ai-programming-...  Not Specified   
3014  https://www.udacity.com/course/data-scientist-...  Not Specified   
3015  https://www.udacity.com/course/front-end-web-d...  Not Specified   
3019  https://www.udacity.com/course/full-stack-web-...  Not Specified   
3020  https://www.udacity.com/course/digital-marketi...  Not Specified   
3022    https://www.udacity.com/course/learn-sql--nd072  Not Specified   
3023  https://www.udacity.com/course/react-nanodegre...  Not Specified   
3024  https://www.udacity.com/course/deep-learning-n...  Not Specified   
3025  https://www.udacity.com/course/ux-designer-nan...  Not Specified   
3027  https://www.udacity.com/course/deep-reinforcem...  Not Specified   
3028  https://www.udacity.com/course/computer-vision...  Not Specified   
3029  https://www.udacity.com/course/java-developer-...  Not Specified   


In [721]:
keyword_map = {
    'machine learning': 'Data Science',
    'data science': 'Data Science',
    'deep learning': 'Data Science',
    'artificial intelligence': 'Data Science',
    'analytics': 'Data Science',
    'sql': 'Data Science',
    'tableau': 'Data Science',
    'data': 'Data Science',
    'analyst': 'Data Science',

    # Computer Science
    'python': 'Computer Science',
    'java': 'Computer Science',
    'c++': 'Computer Science',
    'javascript': 'Computer Science',
    'computer science': 'Computer Science',
    'software engineering': 'Computer Science',
    'web development': 'Computer Science',
    'front-end': 'Computer Science',
    'react': 'Computer Science',

    # Business
    'project management': 'Business',
    'business': 'Business',
    'finance': 'Business',
    'accounting': 'Business',
    'marketing': 'Business',
    'excel': 'Business',

    # Information Technology
    'google cloud': 'Information Technology',
    'aws': 'Information Technology',
    'azure': 'Information Technology',
    'cybersecurity': 'Information Technology',
    'it support': 'Information Technology',
    'scrum': 'Information Technology',
    'data engineering': 'Information Technology',
    'engineer':'Physical Science And Engineering', #general engineer

    # Personal Development
    'leadership': 'Personal Development',
    'career': 'Personal Development',
    'communication': 'Personal Development',

    # Language
    'english': 'Language Learning',
    'spanish': 'Language Learning',
    'speak': 'Language Learning',

    # Other
    'psychology': 'Social Sciences',
    'music': 'Arts and Humanities',
    'health': 'Health'
}

In [722]:
# Step 1: Standardize the 'Master_Category' column to lowercase
df['Master_Category'] = df['Master_Category'].astype(str).str.lower()

In [723]:
# Step 2: Translate the languages
# Use .replace() instead of .map().
# .replace() changes what's in the map and *leaves everything else alone*.
df['Master_Category'] = df['Master_Category'].replace(translation_map)

In [724]:
#Step 3: Capitalize the good data and fix 'nan' / 'other' strings
# We now have 'data science', 'business', 'nan', 'other' etc.
df['Master_Category'] = df['Master_Category'].str.title()
# Turn the string 'Nan' and 'Other' back into real nulls so we can fill them
df['Master_Category'] = df['Master_Category'].replace(['Nan', 'Other'], np.nan)

In [725]:
# Step 4: Fill the *real* nulls using the 'Title' column
# We loop through the keyword_map and fill the nulls.
# This is the correct way to do the keyword search.
for keyword, category in keyword_map.items():

    # Find rows where Category is still null AND Title contains the keyword
    mask = (
        df['Master_Category'].isnull() &
        df['Title'].str.contains(keyword, case=False, na=False)
    )

    # Fill those rows with the correct category
    df.loc[mask, 'Master_Category'] = category

# Step 5: Fill any remaining nulls with 'Other'
df['Master_Category'] = df['Master_Category'].fillna('Other')

In [726]:
print("Category cleaning complete. New counts:")
print(df['Master_Category'].value_counts())

Category cleaning complete. New counts:
Master_Category
Business                            862
Computer Science                    459
Data Science                        457
Health                              251
Information Technology              209
Physical Science And Engineering    144
Arts And Humanities                 120
Language Learning                    86
Social Sciences                      79
Personal Development                 47
Math And Logic                       22
Other                                 1
Name: count, dtype: int64


In [727]:
df.tail(10)

Title  \
3051                            Data Streaming    
3054                          AI for Healthcare   
3055                 Intro to Self-Driving Cars   
3056                 Become a Marketing Analyst   
3057     Intro to Machine Learning with PyTorch   
3058          Predictive Analytics for Business   
3059  Intro to Machine Learning with TensorFlow   
3070  Flying Car and Autonomous Flight Engineer   
3072                         Data Visualization   
3074        Programming for Data Science with R   

                                                    URL    Course Type  \
3051  https://www.udacity.com/course/data-streaming-...  Not Specified   
3054  https://www.udacity.com/course/ai-for-healthca...  Not Specified   
3055  https://www.udacity.com/course/intro-to-self-d...  Not Specified   
3056  https://www.udacity.com/course/marketing-analy...  Not Specified   
3057  https://www.udacity.com/course/intro-to-machin...  Not Specified   
3058  https://www.udacity.com/course/predictive-anal...  Not Specified   
3059  https://www.udacity.com/course/intro-to-machin...  Not Specified   
3070  https://www.udacity.com/course/flying-car-nano...  Not Specified   
3072  https://www.udacity.com/course/data-visualizat...  Not Specified   
3074  https://www.udacity.com/course/programming-for...  Not Specified   

     Language                                             Skills Instructors  \
3051  English   Kafka Rest Proxy, Kafka Connect, Kafka Basics...     Unknown   
3054  English   Feature Engineering, Data Splitting, Model Pe...     Unknown   
3055  English   Python Syntax, Python Data Structures, Python...     Unknown   
3056  English   Filtering in Google Analytics, Google Analyti...     Unknown   
3057  English   Independent Component Analysis, Gaussian Mixt...     Unknown   
3058  English   Univariate Time-Series Analysis, Analytic Fra...     Unknown   
3059  English   Independent Component Analysis, Gaussian Mixt...     Unknown   
3070  English   3D Robot Motion control, PID Controller, Obje...     Unknown   
3072  English   Data Storytelling, Data Visualization Design,...     Unknown   
3074  English   Git, SQL Queries, SQL JOINs, SQL Aggregations...     Unknown   

      Rating  Number of viewers     Site   Master_Category  Duration_Hours  
3051     4.3                0.0  Udacity      Data Science            40.0  
3054     4.6                0.0  Udacity  Computer Science            80.0  
3055     4.6                0.0  Udacity  Computer Science            80.0  
3056     4.6                0.0  Udacity      Data Science            80.0  
3057     4.7                0.0  Udacity      Data Science            60.0  
3058     4.7                0.0  Udacity      Data Science            60.0  
3059     4.7                0.0  Udacity      Data Science            60.0  
3070     4.6                0.0  Udacity  Computer Science            80.0  
3072     4.6                0.0  Udacity      Data Science            80.0  
3074     4.8                0.0  Udacity      Data Science            60.0

In [728]:
#check nulls
df[df['Master_Category'].isnull()]


Empty DataFrame
Columns: [Title, URL, Course Type, Language, Skills, Instructors, Rating, Number of viewers, Site, Master_Category, Duration_Hours]
Index: []

In [729]:
#rename Master
df=df.rename(columns={'Master_Category':'Category'})

In [730]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2737 entries, 0 to 3074
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              2737 non-null   object 
 1   URL                2737 non-null   object 
 2   Course Type        2737 non-null   object 
 3   Language           2737 non-null   object 
 4   Skills             1945 non-null   object 
 5   Instructors        2737 non-null   object 
 6   Rating             2737 non-null   float64
 7   Number of viewers  2737 non-null   float64
 8   Site               2737 non-null   object 
 9   Category           2737 non-null   object 
 10  Duration_Hours     2737 non-null   float64
dtypes: float64(3), object(8)
memory usage: 256.6+ KB


#Skills

The skills are in a list format so we  need to flatten them.For that we will create a diffrent dataframe for skills.But first we will replace nulls and reset the index


In [731]:
# Reset the index and drop the old index
df = df.reset_index(drop=True)

In [732]:
#rename the index column Course_ID
df.index.name = 'Course_ID'

In [733]:
df.tail()

Title  \
Course_ID                                              
2732               Predictive Analytics for Business   
2733       Intro to Machine Learning with TensorFlow   
2734       Flying Car and Autonomous Flight Engineer   
2735                              Data Visualization   
2736             Programming for Data Science with R   

                                                         URL    Course Type  \
Course_ID                                                                     
2732       https://www.udacity.com/course/predictive-anal...  Not Specified   
2733       https://www.udacity.com/course/intro-to-machin...  Not Specified   
2734       https://www.udacity.com/course/flying-car-nano...  Not Specified   
2735       https://www.udacity.com/course/data-visualizat...  Not Specified   
2736       https://www.udacity.com/course/programming-for...  Not Specified   

          Language                                             Skills  \
Course_ID                                                               
2732       English   Univariate Time-Series Analysis, Analytic Fra...   
2733       English   Independent Component Analysis, Gaussian Mixt...   
2734       English   3D Robot Motion control, PID Controller, Obje...   
2735       English   Data Storytelling, Data Visualization Design,...   
2736       English   Git, SQL Queries, SQL JOINs, SQL Aggregations...   

          Instructors  Rating  Number of viewers     Site          Category  \
Course_ID                                                                     
2732          Unknown     4.7                0.0  Udacity      Data Science   
2733          Unknown     4.7                0.0  Udacity      Data Science   
2734          Unknown     4.6                0.0  Udacity  Computer Science   
2735          Unknown     4.6                0.0  Udacity      Data Science   
2736          Unknown     4.8                0.0  Udacity      Data Science   

           Duration_Hours  
Course_ID                  
2732                 60.0  
2733                 60.0  
2734                 80.0  
2735                 80.0  
2736                 60.0

In [734]:
df['Skills'] = df['Skills'].fillna('Not Specified')

In [735]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2737 entries, 0 to 2736
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              2737 non-null   object 
 1   URL                2737 non-null   object 
 2   Course Type        2737 non-null   object 
 3   Language           2737 non-null   object 
 4   Skills             2737 non-null   object 
 5   Instructors        2737 non-null   object 
 6   Rating             2737 non-null   float64
 7   Number of viewers  2737 non-null   float64
 8   Site               2737 non-null   object 
 9   Category           2737 non-null   object 
 10  Duration_Hours     2737 non-null   float64
dtypes: float64(3), object(8)
memory usage: 235.3+ KB


Now our main DataFrame has a Course_ID and no null skills, it's ready to be saved.

The dataset is clean and ready for analysis.

In [736]:
# Save the main file
df.to_excel('cleaned_online_courses.xlsx')

#New dataframe for skills

In [737]:
df_skills = df[['Skills']].copy()
df_skills.head()

Skills
Course_ID                                                   
0          Decision Trees, Artificial Neural Network, Log...
1          Data Science, Relational Database Management S...
2          Data Science, Github, Python Programming, Jupy...
3          Data Science, Artificial Intelligence (AI), Bu...
4          Artificial Neural Network, Convolutional Neura...

In [738]:
df_skills['Skills'] = df_skills['Skills'].str.split(',')
df_skills.head()

Skills
Course_ID                                                   
0          [Decision Trees,  Artificial Neural Network,  ...
1          [Data Science,  Relational Database Management...
2          [Data Science,  Github,  Python Programming,  ...
3          [Data Science,  Artificial Intelligence (AI), ...
4          [Artificial Neural Network,  Convolutional Neu...

In [739]:
df_skills = df_skills.explode('Skills')
df_skills.head(15)

Skills
Course_ID                                                
0                                          Decision Trees
0                               Artificial Neural Network
0                                     Logistic Regression
0                                     Recommender Systems
0                                       Linear Regression
0                     Regularization to Avoid Overfitting
0                                        Gradient Descent
0                                     Supervised Learning
0                  Logistic Regression for Classification
0                                                 Xgboost
0                                              Tensorflow
0                                          Tree Ensembles
0                                                        
1                                            Data Science
1           Relational Database Management System (RDBMS)

In [740]:
df_skills.tail()

Skills
Course_ID                        
2736        R Function Definition
2736            R Data Structures
2736              Data Structures
2736                   UNIX Shell
2736                         IDEs

In [741]:
df_skills['Skills'] = df_skills['Skills'].str.strip()

In [742]:
df_skills = df_skills[df_skills['Skills'] != '']
df_skills = df_skills[df_skills['Skills'] != 'Not Specified']

In [743]:
df_skills.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11521 entries, 0 to 2736
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Skills  11521 non-null  object
dtypes: object(1)
memory usage: 180.0+ KB


In [744]:
# Save the final skills file
df_skills.to_excel('cleaned_course_skills.xlsx')
